hal/lib/athena/SerialHelper.cpp

```diff
@@ -44,66 +46,87 @@ std::string SerialHelper::GetSerialPortName(HAL_SerialPort port,
     return "";
   }
 
-  // Hold lock whenever we're using the names array
-  std::lock_guard<priority_mutex> lock(m_nameMutex);
+  int32_t visaIndex = GetIndexForPort(port, status);
+  if (visaIndex == -1) {
+    *status = HAL_SERIAL_PORT_NOT_FOUND;
+    return "";
+    // Error
+  } else {
+    return m_visaResource[visaIndex].str();
+  }
+}
 
-  std::string portString = m_usbNames[port - 2];
+std::string SerialHelper::GetOSSerialPortName(HAL_SerialPort port,
+                                              int32_t* status) {
+  if (port == HAL_SerialPort::HAL_SerialPort_Onboard) {
+    return OnboardResourceOS;
+  } else if (port == HAL_SerialPort::HAL_SerialPort_MXP) {
+    return MxpResourceOS;
+  }
 
-  llvm::SmallVector<int32_t, 4> indices;
+  QueryHubPaths(status);
 
-  // If port has not been assigned, find the one to assign
-  if (portString.empty()) {
-    for (size_t i = 0; i < 2; i++) {
-      // Remove all used ports
-      auto idx = std::find(m_sortedHubPath.begin(), m_sortedHubPath.end(),
-                           m_usbNames[i]);
-      if (idx != m_sortedHubPath.end()) {
-        // found
-        m_sortedHubPath.erase(idx);
-      }
-      if (m_usbNames[i] == "") {
-        indices.push_back(i);
-      }
-    }
+  // If paths are empty or status error, return error
+  if (*status != 0 || m_visaResource.empty() || m_osResource.empty() ||
+      m_sortedHubPath.empty()) {
+    *status = HAL_SERIAL_PORT_NOT_FOUND;
+    return "";
+  }
 
-    int32_t idx = -1;
-    for (size_t i = 0; i < indices.size(); i++) {
-      if (indices[i] == port - 2) {
-        idx = i;
-        break;
-      }
-    }
+  int32_t osIndex = GetIndexForPort(port, status);
+  if (osIndex == -1) {
+    *status = HAL_SERIAL_PORT_NOT_FOUND;
+    return "";
+    // Error
+  } else {
+    return m_osResource[osIndex].str();
+  }
+}
 
-    if (idx == -1) {
-      *status = HAL_SERIAL_PORT_NOT_FOUND;
-      return "";
-    }
+std::vector<std::string> SerialHelper::GetVISASerialPortList(int32_t* status) {
+  std::vector<std::string> retVec;
 
-    if (idx >= static_cast<int32_t>(m_sortedHubPath.size())) {
-      *status = HAL_SERIAL_PORT_NOT_FOUND;
-      return "";
-    }
+  // Always add 2 onboard ports
+  retVec.push_back(std::string(OnboardResourceVISA));
+  retVec.push_back(std::string(MxpResourceVISA));
```

Comment:
This makes an extra string copy.  Use emplace_back and don't explicitly call the std::string constructor.